text/0012-jsonfile.md

```diff
@@ -0,0 +1,60 @@
+# A brief analysis of JSON file-backed storage
+
+Several components in Firefox — including libpref, XULStore, autofill, and logins — follow a general pattern:
+
+- Store data in memory, usually as a JS collection.
+- Load that data from disk during initialization.
+- Persist that data to disk in its entirety, usually serialized as JSON, at some point after each write, and/or at shutdown.
+
+[JSONFile.jsm](https://dxr.mozilla.org/mozilla-central/source/toolkit/modules/JSONFile.jsm) is a centralized implementation of this pattern. It supports Promise-based asynchronous loading, synchronous loading, and flushes asynchronously, defaulting to 1500ms after each change. libpref flushes after 500msec.
+
+This approach is well suited to data that:
+
+- Changes infrequently.
+- Is relatively small.
+- Is only ever directly read or written by a single thread.
+- Needs to be kept entirely in memory for reads.
+- Has a straightforward JSON encoding (consider the expense of base64 + UTF-8 for images).
```

Comment:
nit: I'd elaborate "consider the expense" to state your conclusion.  Perhaps: "For example, base64 + UTF-8 encoding images is an expensive JSON encoding that is avoided in practice."